### PR TITLE
Create a blog post about cleaning up migrations

### PR DESCRIPTION
**Pull request by bocharsky-bw**

### CODE REVIEW

File: migrations.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
+# Clean up your migrations
```
Comment (weaverryan) [inline migrations.md:1]:
```suggestion
# Clean up your migrations!
```

File: migrations.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+about it, check it out!
+
+But here, I will give you a tip on how to clean up your migrations correctly.
+After a while, you may have a lot of migrations that start running long,
+especially if you're running them on a huge set of data. In this case,
+you can drop all the migrations you have so far and create a single migration
+that will recreate the whole database schema from scratch.
+
+Of course, we're going to do this locally first, so, run:
+```terminal
```
Comment (weaverryan) [inline migrations.md:16]: line break before `terminal`

File: migrations.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+Rename the migration you just created to one that was already executed on
+the production - you can check the latest migration name you dropped by running:
+```terminal
+git status
+```
+
+For example, if you had a migration file called `Version20220415102030.php`
+before - rename the new migration file that was just generated to this exact name,
+and don't forget to change its class name inside the file accordingly. This will
```
Comment (weaverryan) [inline migrations.md:58]:
```suggestion
and don't forget to change the class name inside the file accordingly. This will
```

File: migrations.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+allow you to deploy to production but this migration won't be executed (will be skipped)
```
Comment (weaverryan) [inline migrations.md:59]:
```suggestion
allow you to deploy to production, but this migration won't be executed (it will be skipped)
```

File: migrations.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+because it was already executed in the past.
+
+Basically, that's it. The trick is to *modify* the existing migration our production
+database knows about so that this migration would be skipped because we already have
+a valid schema there.
```
Comment (weaverryan) [inline migrations.md:64]: I think we can remove this paragraph: everything is pretty well-explained already :)

File: migrations.md [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+But we can improve this a little bit more. Well, if you executed this migration
```
Comment (weaverryan) [inline migrations.md:66]: I think the post could use a few `## Headers`. For example, before this, we might have `## Removing the Previously Executed Migrations from Production`

File: migrations.md [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+right now - it would say something like:
+
+> [WARNING] You have X previously executed migrations in the database that are not
+> registered migrations.
+
+every time you execute migrations unless you clean up those deleted migrations from
+the `migration_versions` table. But who wants to remove them manually on production?
```
Comment (weaverryan) [inline migrations.md:73]: Maybe rephrase to:

> This is because the `migrations_versions` table on production will now show that it executed a *bunch*
> of migrations in the past... but these migration files don't exist anymore! We could remove them manually
> from the production database but... yikes! Manually running a `DELETE` query on the production database
> is less fun than taking your cat to the vet.

File: migrations.md [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        // this up() migration is auto-generated, please modify it to your needs
+
+        $this->addSql('DELETE FROM migration_versions WHERE version NOT LIKE "%20220415102030" AND version NOT LIKE "%20220415102031"');
+    }
+    // ...
+}
+```
+
+Where `20220415102031` is the number of the blank migration we just generated,
+and `20220415102030` is the number of another complete migration we have.
```
Comment (weaverryan) [inline migrations.md:95]:
```suggestion
and `20220415102030` is the number of the other complete migration we have.
```

File: migrations.md [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+Make sure you replaced those numbers with yours!
```
Comment (weaverryan) [inline migrations.md:96]:
```suggestion
Make sure you replace those numbers with yours!
```

File: migrations.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+And we're ready! Try it locally first:
+```terminal
+symfony console doctrine:migration:migrate
+```
+
+to make sure it does not fail. It will still show the warning, but only once.
```
Comment (weaverryan) [inline migrations.md:103]:
```suggestion
It will still show the warning, but only once.
```

File: migrations.md [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+All the further runs will not show it anymore. To double-check,  you may want
```
Comment (weaverryan) [inline migrations.md:104]:
```suggestion
To double-check,  you may want
```

File: migrations.md [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+to look at `migration_versions` table in your local database to make sure we
+have only those 2 migrations there.
+
+Now, it's safe to deploy it to production and celebrate with the low total number
```
Comment (weaverryan) [inline migrations.md:108]:
```suggestion
Now, it's safe to deploy it to production and celebrate with a clean `migrations/` directory!
```

File: migrations.md [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+
+Database migrations are a great way to safely update your database schema,
+and so it's very useful for production where you do not want to lose the data
+you have. If. you want to know more about migrations - SymfonyCasts has a
+[screencasts](https://symfonycasts.com/screencast/symfony-doctrine/migrations)
+about it, check it out!
+
+But here, I will give you a tip on how to clean up your migrations correctly.
+After a while, you may have a lot of migrations that start running long,
+especially if you're running them on a huge set of data. In this case,
```
[the remark this answers is not in the document]
Comment (weaverryan) [reply migrations.md:11]: Also make it clear that this is TOTALLY optional. But if a giant `migrations/` folder is bothering you, clean it up! Also, congratulations on having a successful site that has lived long enough to need this ;).

**Review by weaverryan — COMMENTED** [review]
Cool 😎

File: clean-up-migrations.md [hunk 14]
```diff
@@ -0,0 +1,126 @@
+# Clean up your migrations!
+
+Database migrations are a great way to safely update your database schema,
+and so it's very useful for production where you do not want to lose the data
+you have. If. you want to know more about migrations - SymfonyCasts has a
```
Comment (weaverryan) [inline clean-up-migrations.md:5]:
```suggestion
you have. If you want to know more about migrations - we have a
```

File: clean-up-migrations.md [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+[screencasts](https://symfonycasts.com/screencast/symfony-doctrine/migrations)
```
Comment (weaverryan) [inline clean-up-migrations.md:6]:
```suggestion
[screencast](https://symfonycasts.com/screencast/symfony-doctrine/migrations)
```

File: clean-up-migrations.md [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+about it, check it out!
+
+But here, I will give you a tip on how to clean up your migrations correctly.
```
Comment (weaverryan) [inline clean-up-migrations.md:9]:
```suggestion
But here, I want to give you a tip on how to clean up your migrations... if you find your
`migrations/` directory growing and growing... like we did!
```

File: clean-up-migrations.md [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+First of all, this is mostly for local development, because usually it's not
```
Comment (weaverryan) [inline clean-up-migrations.md:10]:
```suggestion
First of all, this is mostly a problem for local development: on production, we don't
really care how many migration files were executed in the past.
```

File: clean-up-migrations.md [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+important how many migrations you have on production - all those migrations
+were already executed.
+
+But in local development it's also a good idea to create the database and run
```
Comment (weaverryan) [inline clean-up-migrations.md:14]:
```suggestion
But while developing locally, to set things up, we recommend creating your database and
then running the `doctrine:migrations:migrate` command so that you have exactly what's
on production, including a `migrations_versions` table. This is better than running the
`doctrine:schema:update` command, which then makes it hard to test new migrations.
```

File: clean-up-migrations.md [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+the migrations so that you have exactly what's on production and a `migration_versions`
+table that matches production instead of simply running `symfony console doctrine:schema:update`
+command.
+
+And that's when many migration files may cause some performance issues locally.
+In this case, you can drop all the migrations you have so far and create a single
+migration that will recreate the whole database schema from scratch.
```
Comment (weaverryan) [inline clean-up-migrations.md:21]: I still don't think we've explained the problem clearly enough. I think we're closer. It should be something like this:

A) Locally, we recommend setting up your local system with `doctrine:migrations:migrate` instead of `doctrine:schema:update` so that you can easily test any new migrations you add.

B) But, if you have a lot of old migration files, this can start to get kinda slow! On Symfonycasts, we had 50 (fill in real number).

C) And so, while it's TOTALLY optional, if a giant `migrations/` folder bothers you, you CAN clean it up.

So, I think just some light rephrasing of these few paragraphs :)

File: clean-up-migrations.md [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+## Generating a Complete Single Migration
+
+Of course, we're going to do this locally first, so, run:
+
+```terminal
+rm -rf migrations/*
+```
+
+to drop all the migrations. Then, make sure you don't have important data
+you care about in your local database and drop its schema completely with:
```
Comment (weaverryan) [inline clean-up-migrations.md:36]:
```suggestion
you care about in your local database, then drop its schema completely with:
```

File: clean-up-migrations.md [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+
+```terminal
+symfony console doctrine:schema:drop --force
+```
+
+***TIP
+The `symfony console` is the same as `bin/console` but allows env vars to be injected
```
Comment (weaverryan) [inline clean-up-migrations.md:43]:
```suggestion
The `symfony console` command is the same as `bin/console` but it allows environment variables to be injected
```

File: clean-up-migrations.md [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+Now let's generate a new migration with
+[MakerBundle](https://symfonycasts.com/screencast/symfony-fundamentals/maker-command):
+
+```terminal
+symfony console make:migration
+```
+
+Double-check the queries inside the new file to make sure the new migration looks good.
+
+But we cannot just commit changes and deploy to production because the database server
```
Comment (weaverryan) [inline clean-up-migrations.md:56]:
```suggestion
But we can't just commit these changes and deploy to production... because the database server
```

File: clean-up-migrations.md [hunk 23]
```diff
@@ -0,0 +1,126 @@
...
+will try to recreate the tables we already have on production and the migration will fail.
```
Comment (weaverryan) [inline clean-up-migrations.md:57]:
```suggestion
will try to run this command and recreate the tables we already have on production!
```

File: clean-up-migrations.md [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+And... there's almost nothing *less* fun than migrations failing during a deploy!
```
Comment (weaverryan) [inline clean-up-migrations.md:58]:
```suggestion
There's almost nothing *less* fun than migrations failing during a deploy!
```

File: clean-up-migrations.md [hunk 25]
```diff
@@ -0,0 +1,126 @@
...
+
+With a trick: rename the migration you just created to one that was already executed on
+production. You can check the latest migration name you dropped by running:
+
+```terminal
+git status
+```
+
+For example, if you had a migration file called `Version20220415102030.php`
+before - rename the new migration file that was just generated to this exact name,
```
Comment (weaverryan) [inline clean-up-migrations.md:71]:
```suggestion
before, rename the new migration file that was just generated to this exact name,
```

File: clean-up-migrations.md [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+and don't forget to change the class name inside the file accordingly. This will
+allow you to deploy to production but this migration won't be executed
```
Comment (weaverryan) [inline clean-up-migrations.md:73]:
```suggestion
allow you to deploy to production, but this migration won't be executed
```

File: clean-up-migrations.md [hunk 27]
```diff
@@ -0,0 +1,126 @@
...
+(it will be skipped) because it was already executed in the past.
+
+## Removing the Previously Executed Migrations from Production
+
+But we can improve this a bit more. Well, if you executed this migration right
```
Comment (weaverryan) [inline clean-up-migrations.md:78]:
```suggestion
But we can improve this a bit more. Because if you executed this migration right
```

File: clean-up-migrations.md [hunk 28]
```diff
@@ -0,0 +1,126 @@
...
+now - it would say something like:
+
+> [WARNING] You have X previously executed migrations in the database that are not
+> registered migrations.
+
+This is because the `migration_versions` table on production will now show that it
```
Comment (weaverryan) [inline clean-up-migrations.md:84]:
```suggestion
Rude! This is because the `migration_versions` table on production will now show that it
```

File: clean-up-migrations.md [hunk 29]
```diff
@@ -0,0 +1,126 @@
...
+Manually running a `DELETE` query on the production database is less fun than taking
+your cat to the vet.
+
+Generate a blank migration instead:
+
+```terminal
+symfony console doctrine:migration:generate
+```
+
+Then, open it and add a new SQL statement at the end:
```
Comment (weaverryan) [inline clean-up-migrations.md:96]:
```suggestion
Then, open it, and add a new SQL statement at the end:
```

File: metadata.yml [hunk 30]
```diff
@@ -333,3 +333,12 @@ blogs:
             Symfony 5 para que puedas disfrutarlo gratuitamente sin tener que activar
             el traductor universal!
         category: tutorials
+
+    "clean-up-migrations":
+        title: 'Clean up your migrations!'
+        published: 2022-04-15
+        preview: |
+            Does your giant migrations/ folder is bothering you? Clean it up! A tip
```
Comment (weaverryan) [inline metadata.yml:341]:
```suggestion
            Does your giant migrations/ folder bother you? Let's do some spring cleaning! A tip
```

File: metadata.yml [hunk 31]
```diff
@@ -333,3 +333,12 @@ blogs:
...
+            about how to get rid of a hude set of old migration files in your project
```
Comment (weaverryan) [inline metadata.yml:342]:
```suggestion
            about how to get rid of a huge set of old migration files in your project, without
            blowing up your database.
```

File: metadata.yml [hunk 32]
```diff
@@ -333,3 +333,12 @@ blogs:
             Symfony 5 para que puedas disfrutarlo gratuitamente sin tener que activar
             el traductor universal!
         category: tutorials
+
+    "clean-up-migrations":
+        title: 'Clean up your migrations!'
+        published: 2022-04-15
```
Comment (weaverryan) [inline metadata.yml:339]: Let's aim for Monday :)

**Comment by bocharsky-bw** [comment]
@weaverryan Thanks a lot! Round 2 comments applied!

**Review by weaverryan — APPROVED** [review]
Approved!
